# Approach 2: Gaussian Filtering on Model 3

In [1]:
from __future__ import print_function
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K
import keras.backend as K
from keras.models import  load_model
import tensorflow as tf
from keras.layers import Input, Reshape, Add
from keras import backend as K
from keras.layers.noise import GaussianNoise
import numpy as np
from functools import reduce
from operator import mul
from keras.models import Model
import tensorflow as tf
import matplotlib.pyplot as plt
from random import randint
import foolbox
import keras
import numpy as np
from random import randint
import cv2
import random

Using TensorFlow backend.


In [2]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

from torchvision import datasets, transforms

import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [3]:
import numpy as np
import matplotlib.pyplot as plt
import cv2
import random, os
import h5py
from sklearn.metrics import accuracy_score



At this step we will define two functions- choose_folder() and image_path() which heps us by, taking two 
random images from the dataset and testing it.

In [4]:
def choose_folder():
    folder_label=["00001","00002","00003","00004","00005","00006","00007",
                   "00008","00009","00010","00011","00012","00013","00014",
                    "00015","00016","00017","00018","00019","00020","00021",
                     "00022","00023","00024","00025","00026","00027","00028",
                     "00029","00030","00031","00032","00033","00034","00035",
                      "00036","00037","00038","00039","00040","00041","00042"]
    
    folder=random.choice(folder_label)
    return folder

def image_path():
    folder=choose_folder()
    print("****************************************************************************")
    print("the true label is",int(folder))
    path = r"/home/pulkit/Downloads/GTSRB/Training/"
    path=path+folder+"/"
    #print("the path is ",path)
    my_image = random.choice([
    x for x in os.listdir(path)
    if os.path.isfile(os.path.join(path, x))
    ])
    p=os.path.join(path, my_image)
    return p,folder

Now, we will define a function which will apply Gausian Blur function with a specified height and width of the 
kernel. The model predicts the label for every image. At last, an accuracy score is calculated.
GaussianBlur is applied 3 times to fool the model, as it is very well trained.

In [6]:
def gausian_filtering_model_3():
    
    y_true=[]
    y_pred=[]
    
    for i in range(0,200,1):
        path,label=image_path()
        y_true.append(int(label))
        img=cv2.imread(path,1)
        
        
        # applying Gausian Bur
        for i in range(0,3):
            img = cv2.GaussianBlur(img,(5,5),0)
        
        # preprocessing the image, so that the model can predict
        
        img= tf.cast(img, tf.float32)/255
        img = tf.image.resize(img, (32, 32))
        img = tf.expand_dims(img , 0)
        
        
        model = tf.keras.models.load_model('/home/pulkit/Desktop/trained/noLoss_model.h5')

        pred=int(model.predict_classes(img))
        print("The predicted label is ",pred)
        y_pred.append(pred)
        
    print("****************************************************************************")        
    correct_percent=accuracy_score(y_true, y_pred)
    print("The accuracy score is ",correct_percent)

In [7]:
gausian_filtering_model_3()

****************************************************************************
the true label is 26
The predicted label is  5
****************************************************************************
the true label is 34
The predicted label is  8
****************************************************************************
the true label is 16
The predicted label is  5
****************************************************************************
the true label is 6
The predicted label is  5
****************************************************************************
the true label is 28
The predicted label is  28
****************************************************************************
the true label is 7
The predicted label is  5
****************************************************************************
the true label is 39
The predicted label is  3
****************************************************************************
the true label is 16
The predicted label is  16
********

The predicted label is  5
****************************************************************************
the true label is 41
The predicted label is  3
****************************************************************************
the true label is 5
The predicted label is  5
****************************************************************************
the true label is 30
The predicted label is  29
****************************************************************************
the true label is 15
The predicted label is  15
****************************************************************************
the true label is 4
The predicted label is  14
****************************************************************************
the true label is 3
The predicted label is  3
****************************************************************************
the true label is 30
The predicted label is  5
****************************************************************************
the true label is 16
The pred

The predicted label is  8
****************************************************************************
the true label is 28
The predicted label is  29
****************************************************************************
the true label is 1
The predicted label is  5
****************************************************************************
the true label is 17
The predicted label is  17
****************************************************************************
the true label is 6
The predicted label is  5
****************************************************************************
the true label is 5
The predicted label is  5
****************************************************************************
the true label is 42
The predicted label is  5
****************************************************************************
the true label is 25
The predicted label is  25
****************************************************************************
the true label is 39
The pred

The predicted label is  5
****************************************************************************
the true label is 38
The predicted label is  38
****************************************************************************
The accuracy score is  0.585


Conclusion-: Applying Gausinan Fitering decreases the prediction accuracy of the model from 93% to 58.5%. 
In order to further decresase the the prediction accuracy, we can apply the Gaussian Filtering once more on
already filtered image. Further applying Gaussian Filtering on the image will make the quality of 
the image so bad, that even the human eye will not be able to recognize it.

Note-: The quality of the image obtained after FGSM attack, is far superior than obtained by Gausian Filter.
 